### ISSUE TITLE
renamed/removed GLIF still lingering after writing to existing UFO path

### ISSUE THREAD

**Issue by MrBrezina** [googlei18n/fontmake#372]
```
Traceback (most recent call last):
  File "/usr/local/bin/fontmake", line 11, in <module>
    load_entry_point('fontmake==1.3.0', 'console_scripts', 'fontmake')()
  File "/usr/local/lib/python2.7/site-packages/fontmake/__main__.py", line 195, in main
    ufo_paths, is_instance=args.pop('masters_as_instances'), **args)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 473, in run_from_ufos
    ufos, remove_overlaps, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 175, in build_otfs
    self.save_otfs(ufos, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
    return func(*args, **kwds)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 315, in save_otfs
    self.subset_otf_from_ufo(otf_path, ufo)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 356, in subset_otf_from_ufo
    subsetter.subset(font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/subset/__init__.py", line 2915, in subset
    self._prune_pre_subset(font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/subset/__init__.py", line 2742, in _prune_pre_subset
    table = font[tag]
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/__init__.py", line 426, in __getitem__
    table.decompile(data, self)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 42, in decompile
    self.table.decompile(reader, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 626, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 506, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 622, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 145, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 506, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 622, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 145, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 506, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 626, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otConverters.py", line 506, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otBase.py", line 631, in decompile
    self.postRead(table, font)
  File "/usr/local/lib/python2.7/site-packages/fontTools/ttLib/tables/otTables.py", line 465, in postRead
    (r.StartCoverageIndex, len(glyphs))
AssertionError: (183, 263)
```

**Comment by anthrotype** [comment]
thanks David. It would be nice to get a copy of the UFO and OTF that trigger this (privately, if you can't share here).

**Comment by anthrotype** [comment]
do you also get any warnings before the assertion error?
in the fonttools code, I see there are several special cases and workarounds when decompiling Coverage tables from broken fonts in the wild:
https://github.com/fonttools/fonttools/blame/master/Lib/fontTools/ttLib/tables/otTables.py#L451-L482

(though admittedly, this OTF itself came from fonttools, via ufo2ft and feaLib).

**Comment by MrBrezina** [comment]
Nope. I will send the UFO privately. I am not getting this error with other fonts, so…

**Comment by anthrotype** [comment]
Thanks!
what is the exact fontmake command line that you used?

**Comment by anthrotype** [comment]
ok reproduced

**Comment by anthrotype** [comment]
ok, making progress.
If you pass `--no-production-names` to fontmake, then it builds.
It may be the renaming of the glyphs that occurs in ufo2ft that somehow leaves the font in a broken state.
investingating

**Comment by anthrotype** [comment]
it looks like the lib.plist postcscriptNames has multiple entries which point to the same glyph name: e.g. both "arFour.URD" and "arFourFarsi.URD" point to "u06F4.URD".
I don't know if there's a requirement for post names or CFF charstrings names to be unique; probably there isn't, although the assumption in our code is that they are.

The UFO spec also doesn't explicitly say that the mappings should be unique, but I believe that was an oversight, because why would you ever want to name two glyphs the same?
http://unifiedfontobject.org/versions/ufo3/lib.plist/

can you try fix that are build again?

I'll try to put a check against that with a nice error message somewhere.

**Comment by MrBrezina** [comment]
Thanks for tracking that down! Indeed that should not have happened. I produce the postscriptNames entries based on the UFO. And guess what happened? We renamed arFour.URD to arFourFarsi.URD, but kept the UFOs in `instances_ufo/`, so the UFOs ended up having both glyphs after conversion from .glyphs to .ufo instances. Well. Lessson learned! Removing pre-existing UFOs from now on to be safe.

**Comment by MrBrezina** [comment]
Ah. wait. It really would be cleaner if fontmake did that because I cannot easily tell what instances a .glyphs file exports to.

d

**Comment by anthrotype** [comment]
how did you generate the ufo instances? With fontmake or with glyphsLib directly?

**Comment by MrBrezina** [comment]
With fontmake.

**Comment by khaledhosny** [comment]
> Removing pre-existing UFOs from now on to be safe.

This have bitten me few times before, IMHO defcon or ufoLib ought to do this automatically.

**Comment by anthrotype** [comment]
The `glyphsLib.build_masters` function used by fontmake always removes the master UFOs before writing them out (it calls `glyphsLib.interpolation.build_designpace` which in turn uses `write_ufo` and `clean_ufo` functions from `glyphsLib.util`; the latter does a `shutil.rmtree` if the master UFO exists).

For instance UFOs, however, fontmake calls `mutatorMath.ufo.build` function, which in turn uses the `DesignSpaceDocumentReader._readSingleInstanceElement`, and the latter does not clean the destination UFO path before calling `instanceObject.save()`.

Perhaps mutatorMath InstanceWriter.save method should call shutil.rmtree if UFO path already exists.

**Comment by anthrotype** [comment]
> We renamed arFour.URD to arFourFarsi.URD, but kept the UFOs in `instances_ufo/`, so the UFOs ended up having both glyphs after conversion from .glyphs to .ufo instances

I still don't understand how this could have have happened though... If one renames a glyph (i.e. changes the name attribute in that glif file, and also changes the entry in the contents.plist file), and then the UFOWriter writes the new UFO with renamed glyph on top of the existing one: the renamed glyph will have a new name, and only the glyph with the new name will be in the contents.plist; even if the old un-renamed glif file were still present in the layer folder (I'd expect it shouldn't be there any more), upon reading the UFO, the glif file which is not referenced in the contents.plist should be ignored, _as if_ it was not present in the UFO, because contents.plist is the ultimate index/mapping of glyph names to glif files, not their mere presence on the filesystem.

We need to replicate and isolate this issue in ufoLib, I guess.
In the meantime, shutil.rmtree before defcon.Font.save() is the workaround.

**Comment by MrBrezina** [comment]
I think I was getting some issues when the glyphOrder did not contain all the glyphs that were in the UFO (typically we leave part components hanging at the end in the Glyphs.app). So I have something like this line in my code:

```
font.lib["public.glyphOrder"] += [g.name for g in font if g.name not in font.lib["public.glyphOrder"]]
```

`font` being a defcon Font object. It is not obvious where defcon gets the list of glyphs, but based on what we discussed it seems to look at the files. So effectively this code is adding the glyphs from not-removed files back.

Perhaps glyphsLib should update the glyphOrder before it generates UFOs?

**Comment by anthrotype** [comment]
what issues? It should be perfectly fine if the public.glyphOrder does not contain all the glyphs (according to UFO spec).
ufo2ft simply place them alphabetically at the end of the ordered glyphs.
note that if you do `for g in font` you don't get a predictable deterministic order; better do `for name in sorted(font.keys()): g = font[name] ...`.

Defcon *ought to* get the list of glyphs from the `contents.plist` only, not from the files.
We need check again this is not a bug in defcon.

GlyphsLib currently always writes a public.glyphOrder according to either the `glyphOrder` custom parameter (plus extra glyphs outside that, in the order in which they appear in the *.glyphs file), or according to the order in which glyphs are stored in the *.glyphs file.

**Comment by MrBrezina** [comment]
It was not about being deterministic. It really was just to avoid the issue.

ok, I will stop doing this now and let’s see where that leads.

**Comment by MrBrezina** [comment]
So, did a bit of testing and:

- changing my code, did not change anything. Removed glyphs kept on creeping in,
- making sure the glyphOrder custom parameter did NOT include the recently removed glyphs helped and they stayed removed even though the files were in the instance UFOs.

So I guess glyphsLib should subset the glyphOrder to existing glyphs only (because there are many people who keep a bigger glyphOrder for various reasons).

Hope this helps.

**Comment by anthrotype** [comment]
sorry i'm a bit confused (too many things flying in my head), could you rephrase what's the issue is here? :)

> Removed glyphs kept on creeping in

I need exact steps to reproduce this.

> glyphsLib should subset the glyphOrder to existing glyphs only

hmm... you're saying that if glyphOrder contains glyphs that are no longer in contents.plist but their file still is in the layer sub-folder, defcon believes that they are still there? This doesn't make sense to me

**Comment by MrBrezina** [comment]
ok, the story is this: I have a .glyphs source file with glyph X and this glyph in in the custom parameter glyphOrder. Then I run a workflow which uses fontmake to export UFO instances from the .glyphs file. I add some features using my scripts and then compile final OTFs from the UFO instances. Then I decide to remove X. I delete it from the .glyphs file, but forget to update the glyphOrder. I run the workflow again (not deleting the `instance_ufo/` and `master_ufo/` folders fontmake creates and which by this time still contain the file for the glyph, i.e. X.glif) and compile and the glyph is still there.

If I do not forget to remove it from the custom parameter glyphOrder, it stays removed even though the X.glif file is still in `instance_ufo/`.

> hmm... you're saying that if glyphOrder contains glyphs that are no longer in contents.plist but their file still is in the layer sub-folder, defcon believes that they are still there? This doesn't make sense to me

I am not entirely sure which library does this (defcon was used in my code above, but that bit of code was not causing the problem as I have found in the end), but it seems like something somewhere is using glyphOrder as a list of glyphs that exist in the UFO.

**Comment by anthrotype** [comment]
Thanks for the clarification.
The instance_ufo are being saved by MutatorMath, via defcon, via ufoLib.
Ultimately, ufoLib UFOWriter (called by defcon.Font.save()) should remove any lingering .glif files if they are no longer referenced in the contents.plist. If this doesn't happen when saving to the same location as a pre-existing UFO, then I'd say this is bug.
The glyphOrder does not have anything to do with "what glyph is present?", just with "in which order should I compile (or display) the glyphs that are actually present?". At least, in theory.

**Comment by anthrotype** [comment]
> I decide to remove X. I delete it from the .glyphs file, but forget to update the glyphOrder. I run the workflow again (not deleting the instance_ufo/ and master_ufo/ folders fontmake creates and which by this time still contain the file for the glyph, i.e. X.glif) and compile and the glyph is still there.
> If I do not forget to remove it from the custom parameter glyphOrder, it stays removed even though the X.glif file is still in instance_ufo/...
> it seems like something somewhere is using glyphOrder as a list of glyphs that exist in the UFO

Removing the "predicted" UFO instances before generating new ones with mutatorMath (6cdf44e) is too much of a hack.
I think we need to properly track this issue down to where it belongs, i.e. either defcon and/or ufoLib.

I think it's reasonable to:

1) want to open a UFO, modify something and save it back to the original location, without needing to wipe out the whole UFO
2) saving a different UFO (created from scratch, or loaded from a different path) to the location of an existing UFO, in which case I expect the entire directory to be removed before saving the new one.
